mac/mcompile/mcompile.cpp

```diff
@@ -0,0 +1,638 @@
+  KMX_WCHAR dkid = 0;
+  while (str && *str) {
+    if (*str == UC_SENTINEL && *(str + 1) == CODE_DEADKEY) {
+      dkid = std::max(dkid, *(str + 2));
+    }
+    str = KMX_incxstr(str);
+  }
+  return dkid;
+}
+
+struct KMX_dkidmap {
+  KMX_WCHAR src_deadkey, dst_deadkey;
+};
+
+/**
+ * @brief  find the deadkey id for a given deadkey
+ * @param  kbd     pointer to the keyboard
+ * @param  deadkey for which an id is to be found
+ * @return 0 if failed;
+ *         otherwise a deadkey-id
+ */
+KMX_WCHAR mac_KMX_GetUniqueDeadkeyID(LPKMX_KEYBOARD kbd, KMX_WCHAR deadkey) {
+  LPKMX_GROUP gp;
+  LPKMX_KEY kp;
+  LPKMX_STORE sp;
+  KMX_DWORD i, j;
+  KMX_WCHAR dkid = 0;
+  static KMX_WCHAR s_next_dkid = 0;
+  static KMX_dkidmap* s_dkids = NULL;
+  static int s_ndkids = 0;
+
+  if (!kbd) {
+    if (s_dkids) {
+      delete s_dkids;
+    }
+    s_dkids = NULL;
+    s_ndkids = 0;
+    s_next_dkid = 0;
+    return 0;
+  }
+
+  for (int xi = 0; xi < s_ndkids; xi++) {
+    if (s_dkids[xi].src_deadkey == deadkey) {
+      return s_dkids[xi].dst_deadkey;
+    }
+  }
+
+  if (s_next_dkid != 0) {
+    s_dkids = (KMX_dkidmap*)realloc(s_dkids, sizeof(KMX_dkidmap) * (s_ndkids + 1));
+    s_dkids[s_ndkids].src_deadkey = deadkey;
+    return s_dkids[s_ndkids++].dst_deadkey = ++s_next_dkid;
+  }
+
+  for (i = 0, gp = kbd->dpGroupArray; i < kbd->cxGroupArray; i++, gp++) {
+    for (j = 0, kp = gp->dpKeyArray; j < gp->cxKeyArray; j++, kp++) {
+      dkid = std::max(dkid, mac_KMX_ScanXStringForMaxDeadkeyID(kp->dpContext));
+      dkid = std::max(dkid, mac_KMX_ScanXStringForMaxDeadkeyID(kp->dpOutput));
+    }
+    dkid = std::max(dkid, mac_KMX_ScanXStringForMaxDeadkeyID(gp->dpMatch));
+    dkid = std::max(dkid, mac_KMX_ScanXStringForMaxDeadkeyID(gp->dpNoMatch));
+  }
+
+  for (i = 0, sp = kbd->dpStoreArray; i < kbd->cxStoreArray; i++, sp++) {
+    dkid = std::max(dkid, mac_KMX_ScanXStringForMaxDeadkeyID(sp->dpString));
+  }
+
+  s_dkids = (KMX_dkidmap*)realloc(s_dkids, sizeof(KMX_dkidmap) * (s_ndkids + 1));
+  s_dkids[s_ndkids].src_deadkey = deadkey;
+  return s_dkids[s_ndkids++].dst_deadkey = s_next_dkid = ++dkid;
+}
+
+/**
+ * @brief  Lookup the deadkey table for the deadkey in the physical keyboard. Then for each character, go through and map it through
+ * @param  kbd        pointer to the keyboard
+ * @param  vk_US      virtual key of the us keyboard
+ * @param  shift      shiftstate
+ * @param  deadkey    character produced by a deadkey
+ * @param  all_vector vector that holds the data of the US keyboard and the currently used (underlying) keyboard
+ * @param  keymap     pointer to the currently used (underlying) keyboard Layout
+ * @param  dk_Table   a vector of all possible deadkey combinations for all Linux keyboards
+ */
+void mac_KMX_ConvertDeadkey(LPKMX_KEYBOARD kbd, KMX_WORD vk_US, KMX_DWORD shift, KMX_WCHAR deadkey, vec_dword_3D& all_vector, const UCKeyboardLayout* keyboard_layout) {
+  KMX_WORD deadkeys[512] = {0};
```

Comment:
- changed to 1000 dk-pairs (=3000+3 elements in dk array) Do we need more?
- in case of overflow: 
  - stop processing more dk and 
  - write out a warning ("x deadkeys have not been processed")

Reply:
Better to dynamically allocate I think. Just use a `std::vector<KMX_WORD>` and dynamically expand

Reply:
OK changed